docs/source/notes/resumption.rst

```diff
@@ -0,0 +1,134 @@
+
+    trainer = Trainer(
+        model=model,
+        train_dataloader=train_dataloader,
+        max_duration="1ep",
+        save_filename="ep{epoch}.pt",
+        save_folder="./path/to/checkpoints",
+        save_overwrite=True,
+        save_interval="1ep",  # Save checkpoints every epoch
+    )
+    trainer.fit()
+
+    assert os.path.exists("./path/to/checkpoints/ep1.pt")
+
+    if not os.path.exists("./path/to/checkpoints/ep25.pt"):
+        shutil.copy2("./path/to/checkpoints/ep1.pt", "./path/to/checkpoints/ep25.pt")
+
+    assert os.path.exists("./path/to/checkpoints/ep25.pt")
+
+
+.. testcode::
+
+    x = 2
+
+However, recovering from a failure here would still require manual intervention to relaunch a new job with the ``load_path`` pointing to the correct checkpoint.
+
+Instead, our trainer supports the ``autoresume=True`` feature. With autoresume, the trainer will automatically check the ``save_folder`` for the latest checkpoints and resume training.
+
+.. testcode::
+
+    trainer = Trainer(
+        ...,
+        autoresume=True,
+        save_folder='./path/to/folder',
+        run_name='my_cool_run',
+    )
+
+With autoresume, users can re-submit the _same_ code to the training run, and the trainer will handle finding and resuming from the latest checkpoints. This works well with systems like Kubernetes that automatically resubmit the same job when there is a node failure (due to spot instances as well). For ``autoresume=True`` to work, we require that both a ``save_folder`` and a ``run_name`` be provided. These are used to search for existing checkpoints.
+
+For an example code, see ``examples/checkpoint_autoresume.py``.
+
+Implementation
+--------------
+
+During training, the trainer always symlinks the latest checkpoint to a format (default is ``latest-rank{rank}`` for local files and ``{run_name}/checkpoints/latest-rank{rank}`` for object stores). When ``autoresume=True``, the Trainer searches for checkpoints of that format in the following order:
+
+1. Local checkpoints of the format ``"{save_folder}/latest-rank0"``. The format for the latest checkpoint can be configured with ``save_latest_filename`` argument (default: ``latest-rank{rank}``).
+2. If no local checkpoints are found, then each logger is checked for files of the format ``"{run_name}/checkpoints/latest-rank{rank}"``. This is often used for resuming from an object store such as S3.
+3. Finally, ``load_path`` is used to load a checkpoint. This can be used for example, a fine-tuning run on a spot instance, where ``load_path`` would be set to the original weights.
+
+Example: Object Store
+---------------------
+
+A typical use case is saving checkpoints to object store (e.g. S3) when there is no local file storage shared across runs. For example, a setup such as this:
+
+
+.. testcode::
+
+    from composer.loggers import ObjectStoreLogger
+
+    object_store_logger = ObjectStoreLogger(
+        object_store_cls=LibcloudObjectStore,
```

Comment:
For S3, would it be possible to use the S3ObjectStore class? That will automatically load creds from the environment.

Reply:
Done